### DISCUSSION TITLE
How to check decimal places

### DISCUSSION THREAD

**Discussion by theseeker58** [victools/jsonschema-generator#454] [Q&A] [answered]
I need to generate a json schema starting from a java class that have a property like this

```java
    @XmlValue
    @DecimalMin(value = "0.01", inclusive = true)
    @Digits(integer = 18, fraction = 2)
    protected BigDecimal value;
```

the tool generates this

```json
        "value" : {
          "type" : "number",
          "minimum" : 0.01
        }
```

Please who knows how can the fraction be checked?
TIA

**Comment by CarstenWickner** [comment, marked as answer]
Hi @theseeker58,

The fraction could be enforced via the `multipleOf` keyword (if you don't mind possible trailing zeros), which you can set like this (refer to [documentation](https://victools.github.io/jsonschema-generator/#quot-multipleof-quot-keyword)):
```java
configBuilder.forFields().withNumberMultipleOfResolver(field -> Optional
            .ofNullable(field.getAnnotationConsideringFieldAndGetterIfSupported(Digits.class))
            .map(annotation -> BigDecimal.ONE.divide(BigDecimal.TEN.pow(annotation.fraction()), annotation.fraction(), RoundingMode.FLOOR)).orElse(null));
```
The maximum number of digits before the decimal separator could similarly be enforced as an `exclusiveMaximum` (refer to [documentation](https://victools.github.io/jsonschema-generator/#quot-exclusivemaximum-quot-keyword)) like this:
```java
configBuilder.forFields().withNumberExclusiveMaximumResolver(field -> Optional
            .ofNullable(field.getAnnotationConsideringFieldAndGetterIfSupported(Digits.class))
            .map(annotation -> BigDecimal.TEN.pow(1 + annotation.integer())).orElse(null));
```

That should produce output like the following:
```json
        "value" : {
          "type" : "number",
          "minimum" : 0.01,
          "exclusiveMaximum": 1000000000000000000,
          "multipleOf": 0.01
        }
```

Beyond these, I don't think the [JSON Schema Specification](https://json-schema.org/specification) has particular keywords for the allowed number of digits before and after the decimal separator.

**Reply by theseeker58** [reply]
Thank you @CarstenWickner for your quick reply,
Since I use the library as maven plugin, I guess I have to implement the Module interface, right?
Another question please. Is it possible to apply the restriction only for specific field/fields? This one contains Euros. Another field can contain other currencies up to 5 decimal digits.
TIA

**Comment by theseeker58** [comment]
Thank you @CarstenWickner 
I have implemented the Module interface and everything works right. The only change I made is the use of the overloaded pow method that accepts a MathContext, otherwise I was getting an ArithmeticException, Invalid operation, due to a negative exponent.
Please, let me know if the fraction can be enforced only for a specific field.
Cheers

**Reply by CarstenWickner** [reply]
I've adjusted my proposal above to avoid the negative input to the `pow()` method.
That should hopefully do it.

**Reply by theseeker58** [reply]
Sorry, my fault,
actually I realised that through Digits.class the fraction is dinamically read from the bean balidation, therefore it's perfect in that way.
Cheers
